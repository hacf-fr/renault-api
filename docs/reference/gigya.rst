Gigya low-level authentication API Reference
============================================

.. contents::
    :local:
    :backlinks: none


Core
----

.. automodule:: renault_api.gigya
   :members:

Data models
-----------

.. automodule:: renault_api.gigya.models
   :members:
