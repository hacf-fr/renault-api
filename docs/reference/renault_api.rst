Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
