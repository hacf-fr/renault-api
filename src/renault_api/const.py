"""Constants for Renault API."""
CONF_COUNTRY = "country"
CONF_LOCALE = "locale"
CONF_GIGYA_APIKEY = "gigya-api-key"
CONF_GIGYA_URL = "gigya-root-url"
CONF_KAMEREON_APIKEY = "kamereon-api-key"
CONF_KAMEREON_URL = "kamereon-root-url"

PERMANENT_KEYS = [
    CONF_COUNTRY,
    CONF_LOCALE,
    CONF_GIGYA_APIKEY,
    CONF_GIGYA_URL,
    CONF_KAMEREON_APIKEY,
    CONF_KAMEREON_URL,
]


GIGYA_URL_EU = "https://accounts.eu1.gigya.com"
GIGYA_URL_US = "https://accounts.us1.gigya.com"
KAMEREON_APIKEY = "Ae9FDWugRxZQAGm3Sxgk7uJn6Q4CGEA2"
KAMEREON_URL_EU = "https://api-wired-prod-1-euw1.wrd-aws.com"
KAMEREON_URL_US = "https://api-wired-prod-1-usw2.wrd-aws.com"

LOCALE_BASE_URL = (
    "https://renault-wrd-prod-1-euw1-myrapp-one.s3-eu-west-1.amazonaws.com"
)

AVAILABLE_LOCALES = {
    "bg_BG": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3__3ER_6lFvXEXHTP_faLtq6eEdbKDXd9F5GoKwzRyZq37ZQ-db7mXcLzR1Jtls5sn",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "cs_CZ": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_oRlKr5PCVL_sPWUZdJ8c5NOl5Ej8nIZw7VKG7S9Rg36UkDszFzfHfxCaUAUU5or2",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "da_DK": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_5x-2C8b1R4MJPQXkwTPdIqgBpcw653Dakw_ZaEneQRkTBdg9UW9Qg_5G-tMNrTMc",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "de_DE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_7PLksOyBRkHv126x5WhHb-5pqC1qFR8pQjxSeLB6nhAnPERTUlwnYoznHSxwX668",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "de_AT": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3__B4KghyeUb0GlpU62ZXKrjSfb7CPzwBS368wioftJUL5qXE0Z_sSy0rX69klXuHy",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "de_CH": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_UyiWZs_1UXYCUqK_1n7l7l44UiI_9N9hqwtREV0-UYA_5X7tOV-VKvnGxPBww4q2",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "en_GB": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_e8d4g4SE_Fo8ahyHwwP7ohLGZ79HKNN2T8NjQqoNnk6Epj6ilyYwKdHUyCw3wuxz",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "en_IE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_Xn7tuOnT9raLEXuwSI1_sFFZNEJhSD0lv3gxkwFtGI-RY4AgiePBiJ9EODh8d9yo",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "es_ES": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_DyMiOwEaxLcPdBTu63Gv3hlhvLaLbW3ufvjHLeuU8U5bx3zx19t5rEKq7KMwk9f1",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "es_MX": {
        CONF_GIGYA_URL: GIGYA_URL_US,
        CONF_GIGYA_APIKEY: "3_BFzR-2wfhMhUs5OCy3R8U8IiQcHS-81vF8bteSe8eFrboMTjEWzbf4pY1aHQ7cW0",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_US,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fi_FI": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_xSRCLDYhk1SwSeYQLI3DmA8t-etfAfu5un51fws125ANOBZHgh8Lcc4ReWSwaqNY",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_FR": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_4LKbCcMMcvjDm3X89LU4z4mNKYKdl_W0oD9w-Jvih21WqgJKtFZAnb9YdUgWT9_a",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_BE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_ZK9x38N8pzEvdiG7ojWHeOAAej43APkeJ5Av6VbTkeoOWR4sdkRc-wyF72HzUB8X",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_CH": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_h3LOcrKZ9mTXxMI9clb2R1VGAWPke6jMNqMw4yYLz4N7PGjYyD0hqRgIFAIHusSn",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "fr_LU": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_zt44Wl_wT9mnqn-BHrR19PvXj3wYRPQKLcPbGWawlatFR837KdxSZZStbBTDaqnb",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "hr_HR": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_HcDC5GGZ89NMP1jORLhYNNCcXt7M3thhZ85eGrcQaM2pRwrgrzcIRWEYi_36cFj9",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "hu_HU": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_nGDWrkSGZovhnVFv5hdIxyuuCuJGZfNmlRGp7-5kEn9yb0bfIfJqoDa2opHOd3Mu",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "it_IT": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_js8th3jdmCWV86fKR3SXQWvXGKbHoWFv8NAgRbH7FnIBsi_XvCpN_rtLcI07uNuq",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "it_CH": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_gHkmHaGACxSLKXqD_uDDx415zdTw7w8HXAFyvh0qIP0WxnHPMF2B9K_nREJVSkGq",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "nl_NL": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_ZIOtjqmP0zaHdEnPK7h1xPuBYgtcOyUxbsTY8Gw31Fzy7i7Ltjfm-hhPh23fpHT5",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "nl_BE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_yachztWczt6i1pIMhLIH9UA6DXK6vXXuCDmcsoA4PYR0g35RvLPDbp49YribFdpC",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "no_NO": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_QrPkEJr69l7rHkdCVls0owC80BB4CGz5xw_b0gBSNdn3pL04wzMBkcwtbeKdl1g9",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "pl_PL": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_2YBjydYRd1shr6bsZdrvA9z7owvSg3W5RHDYDp6AlatXw9hqx7nVoanRn8YGsBN8",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "pt_PT": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3__afxovspi2-Ip1E5kNsAgc4_35lpLAKCF6bq4_xXj2I2bFPjIWxAOAQJlIkreKTD",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "ro_RO": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_WlBp06vVHuHZhiDLIehF8gchqbfegDJADPQ2MtEsrc8dWVuESf2JCITRo5I2CIxs",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "ru_RU": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_N_ecy4iDyoRtX8v5xOxewwZLKXBjRgrEIv85XxI0KJk8AAdYhJIi17LWb086tGXR",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "sk_SK": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_e8d4g4SE_Fo8ahyHwwP7ohLGZ79HKNN2T8NjQqoNnk6Epj6ilyYwKdHUyCw3wuxz",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "sl_SI": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_QKt0ADYxIhgcje4F3fj9oVidHsx3JIIk-GThhdyMMQi8AJR0QoHdA62YArVjbZCt",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
    "sv_SE": {
        CONF_GIGYA_URL: GIGYA_URL_EU,
        CONF_GIGYA_APIKEY: "3_EN5Hcnwanu9_Dqot1v1Aky1YelT5QqG4TxveO0EgKFWZYu03WkeB9FKuKKIWUXIS",  # noqa
        CONF_KAMEREON_URL: KAMEREON_URL_EU,
        CONF_KAMEREON_APIKEY: KAMEREON_APIKEY,
    },
}
