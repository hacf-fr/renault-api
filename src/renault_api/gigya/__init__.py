"""Gigya API."""
import logging
from typing import Any
from typing import cast
from typing import Dict

import aiohttp
from marshmallow.schema import Schema

from . import models
from . import schemas

GIGYA_JWT = "gigya_jwt"
GIGYA_LOGIN_TOKEN = "gigya_login_token"
GIGYA_PERSON_ID = "gigya_person_id"
GIGYA_KEYS = [GIGYA_JWT, GIGYA_LOGIN_TOKEN, GIGYA_PERSON_ID]

_LOGGER = logging.getLogger(__name__)


async def request(
    websession: aiohttp.ClientSession,
    method: str,
    url: str,
    data: Dict[str, Any],
    schema: Schema,
) -> models.GigyaResponse:
    """Send request to Gigya."""
    async with websession.request(method, url, data=data) as http_response:
        response_text = await http_response.text()
        # Disable logging on Gigya, to avoid unnecessary exposure.
        # _LOGGER.debug(
        #    "Received Gigya response %s on %s: %s",
        #    http_response.status,
        #    url,
        #    response_text,
        # )
        gigya_response: models.GigyaResponse = schema.loads(response_text)
        # Check for Gigya error
        gigya_response.raise_for_error_code()
        # Check for HTTP error
        http_response.raise_for_status()

        return gigya_response


async def login(
    websession: aiohttp.ClientSession,
    root_url: str,
    api_key: str,
    login_id: str,
    password: str,
) -> models.GigyaLoginResponse:
    """Send POST to /accounts.login."""
    return cast(
        models.GigyaLoginResponse,
        await request(
            websession,
            "POST",
            f"{root_url}/accounts.login",
            data={
                "ApiKey": api_key,
                "loginID": login_id,
                "password": password,
            },
            schema=schemas.GigyaLoginResponseSchema,
        ),
    )


async def get_account_info(
    websession: aiohttp.ClientSession,
    root_url: str,
    api_key: str,
    login_token: str,
) -> models.GigyaGetAccountInfoResponse:
    """Send POST to /accounts.getAccountInfo."""
    return cast(
        models.GigyaGetAccountInfoResponse,
        await request(
            websession,
            "POST",
            f"{root_url}/accounts.getAccountInfo",
            data={
                "ApiKey": api_key,
                "login_token": login_token,
            },
            schema=schemas.GigyaGetAccountInfoResponseSchema,
        ),
    )


async def get_jwt(
    websession: aiohttp.ClientSession,
    root_url: str,
    api_key: str,
    login_token: str,
) -> models.GigyaGetJWTResponse:
    """Send POST to /accounts.getJWT."""
    return cast(
        models.GigyaGetJWTResponse,
        await request(
            websession,
            "POST",
            f"{root_url}/accounts.getJWT",
            data={
                "ApiKey": api_key,
                "login_token": login_token,
                # gigyaDataCenter may be needed for future jwt validation
                "fields": "data.personId,data.gigyaDataCenter",
                "expiration": 900,
            },
            schema=schemas.GigyaGetJWTResponseSchema,
        ),
    )
