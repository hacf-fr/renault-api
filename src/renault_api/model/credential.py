"""Credential dataclass."""
import time
from dataclasses import dataclass

import jwt


@dataclass
class Credential:
    """Base credential that never expires."""

    value: str

    def has_expired(self) -> bool:
        """Check if Credential has expired."""
        return False


@dataclass
class JWTCredential(Credential):
    """JWT credential with expiry time."""

    expiry: float

    def __init__(self, value: str) -> None:
        """Initialise JWTCredential."""
        decoded_token = jwt.decode(value, options={"verify_signature": False})

        super().__init__(value=value)
        self.expiry = decoded_token["exp"]

    def has_expired(self) -> bool:
        """Check if JWT token has expired."""
        return self.expiry < time.time()
