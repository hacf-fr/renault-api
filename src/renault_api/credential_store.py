"""Kamereon client for interaction with Renault servers."""
import json
import os
from typing import Dict
from typing import List
from typing import Optional

import jwt

from renault_api.const import PERMANENT_KEYS
from renault_api.credential import Credential
from renault_api.credential import JWTCredential


class CredentialStore:
    """Credential store."""

    def __init__(self) -> None:
        """Initialise the credential store."""
        self._store: Dict[str, Credential] = {}

    def __getitem__(self, name: str) -> Credential:
        """Get a credential the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred
        raise KeyError(name)

    def get(self, name: str) -> Optional[Credential]:
        """Get a credential the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred
        return None

    def get_value(self, name: str) -> Optional[str]:
        """Get a credential value from the credential store."""
        if name in list(self._store.keys()):
            cred = self._store[name]
            if not cred.has_expired():
                return cred.value
        return None

    def __delitem__(self, name: str) -> None:
        """Remove a credential from the credential store."""
        del self._store[name]
        self._write()

    def __setitem__(self, name: str, value: Credential) -> None:
        """Add a credential to the credential store."""
        if not isinstance(name, str):  # pragma: no cover
            raise TypeError("`name` must be a string")
        if not isinstance(value, Credential):  # pragma: no cover
            raise TypeError("`value` must be a Credential")

        self._store[name] = value
        self._write()

    def __contains__(self, name: str) -> bool:
        """Check if a credential is in the credential store."""
        if name in self._store:
            cred = self._store[name]
            if not cred.has_expired():
                return True
        return False

    def _write(self) -> None:
        """Writes the content to fixed storage."""
        pass

    def clear(self) -> None:
        """Remove all non-permanent keys from credential store."""
        for key in list(self._store.keys()):
            if key not in PERMANENT_KEYS:
                del self._store[key]
        self._write()

    def clear_keys(self, to_delete: List[str]) -> None:
        """Remove specified keys from credential store."""
        for key in list(self._store.keys()):
            if key in to_delete:
                del self._store[key]
        self._write()


class CredentialEncoder(json.JSONEncoder):
    """Custom encoder for Credential class."""

    def default(self, obj: Credential) -> str:
        """Store the value."""
        return obj.value


class FileCredentialStore(CredentialStore):
    """Credential store with items stored in a file."""

    def __init__(self, store_location: str) -> None:
        """Initialise the credential store."""
        super().__init__()
        self._store_location = store_location
        self._read()

    def _read(self) -> None:
        """Read data from store location."""
        if not os.path.exists(self._store_location):
            return
        with open(self._store_location) as json_file:
            data = json.load(json_file)
            for key, value in data.items():
                if key == "gigya_jwt":
                    try:
                        self[key] = JWTCredential(value)
                    except jwt.ExpiredSignatureError:  # pragma: no cover
                        pass
                else:
                    self[key] = Credential(value)

    def _write(self) -> None:
        """Write data to store location."""
        dirname = os.path.dirname(self._store_location)
        if not os.path.isdir(dirname):
            os.makedirs(dirname)
        with open(self._store_location, "w") as json_file:
            json.dump(self._store, json_file, cls=CredentialEncoder)
