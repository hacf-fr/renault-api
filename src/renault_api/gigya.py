"""Gigya client for authentication."""
import logging
from typing import Any
from typing import cast
from typing import Dict

from aiohttp import ClientSession
from marshmallow.schema import Schema

from .const import CONF_GIGYA_APIKEY
from .const import CONF_GIGYA_URL
from renault_api.model import gigya as model

_LOGGER = logging.getLogger(__name__)


class Gigya:
    """Gigya client for authentication."""

    def __init__(
        self, websession: ClientSession, locale_details: Dict[str, str]
    ) -> None:
        """Initialise Gigya."""
        self._websession = websession

        self._api_key = locale_details[CONF_GIGYA_APIKEY]
        self._root_url = locale_details[CONF_GIGYA_URL]

    async def _post(
        self, path: str, data: Dict[str, Any], schema: Schema
    ) -> model.GigyaResponse:
        url = f"{self._root_url}{path}"
        async with self._websession.request("POST", url, data=data) as http_response:
            response_text = await http_response.text()
            _LOGGER.debug(
                "Received Gigya response %s on %s: %s",
                http_response.status,
                url,
                response_text,
            )
            gigya_response: model.GigyaResponse = schema.loads(response_text)
            # Check for Gigya error
            gigya_response.raise_for_error_code()
            # Check for HTTP error
            http_response.raise_for_status()

            return gigya_response

    async def login(self, login_id: str, password: str) -> model.GigyaLoginResponse:
        """POST to /accounts.login."""
        return cast(
            model.GigyaLoginResponse,
            await self._post(
                "/accounts.login",
                data={
                    "ApiKey": self._api_key,
                    "loginID": login_id,
                    "password": password,
                },
                schema=model.GigyaLoginResponseSchema,
            ),
        )

    async def get_account_info(
        self, login_token: str
    ) -> model.GigyaGetAccountInfoResponse:
        """POST to /accounts.getAccountInfo."""
        return cast(
            model.GigyaGetAccountInfoResponse,
            await self._post(
                "/accounts.getAccountInfo",
                data={"ApiKey": self._api_key, "login_token": login_token},
                schema=model.GigyaGetAccountInfoResponseSchema,
            ),
        )

    async def get_jwt(self, login_token: str) -> model.GigyaGetJWTResponse:
        """POST to /accounts.getAccountInfo."""
        return cast(
            model.GigyaGetJWTResponse,
            await self._post(
                "/accounts.getJWT",
                data={
                    "ApiKey": self._api_key,
                    "login_token": login_token,
                    # gigyaDataCenter may be needed for future jwt validation
                    "fields": "data.personId,data.gigyaDataCenter",
                    "expiration": 900,
                },
                schema=model.GigyaGetJWTResponseSchema,
            ),
        )
