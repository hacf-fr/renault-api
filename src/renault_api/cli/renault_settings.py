"""Singletons for the CLI."""
import os
from textwrap import TextWrapper
from typing import Any
from typing import Dict
from typing import Optional

import aiohttp
import click
from tabulate import tabulate

from renault_api.const import CONF_LOCALE
from renault_api.credential import Credential
from renault_api.credential_store import CredentialStore
from renault_api.helpers import get_api_keys

CONF_ACCOUNT_ID = "accound-id"
CONF_VIN = "vin"

CREDENTIAL_PATH = "~/.credentials/renault-api.json"


async def set_options(
    websession: aiohttp.ClientSession,
    ctx_data: Dict[str, Any],
    locale: Optional[str],
    account: Optional[str],
    vin: Optional[str],
) -> None:
    """Set configuration keys."""
    credential_store: CredentialStore = ctx_data["credential_store"]
    if locale:
        # Ensure API keys are available
        api_keys = await get_api_keys(locale, websession=websession)

        credential_store[CONF_LOCALE] = Credential(locale)
        for k, v in api_keys.items():
            credential_store[k] = Credential(v)

    if account:
        credential_store[CONF_ACCOUNT_ID] = Credential(account)
    if vin:
        credential_store[CONF_VIN] = Credential(vin)


def display_settings(ctx_data: Dict[str, Any]) -> None:
    """Get the current configuration keys."""
    credential_store: CredentialStore = ctx_data["credential_store"]
    wrapper = TextWrapper(width=80)
    items = list(
        [key, "\n".join(wrapper.wrap(credential_store.get_value(key) or "-"))]
        for key in credential_store._store.keys()
    )
    click.echo(tabulate(items, headers=["Key", "Value"]))


def reset() -> None:
    """Clear all credentials/settings from the credential store."""
    try:
        os.remove(os.path.expanduser(CREDENTIAL_PATH))
    except FileNotFoundError:
        pass
