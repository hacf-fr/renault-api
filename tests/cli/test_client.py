"""Test cases for the __main__ module."""
import os
from locale import getdefaultlocale

from aioresponses import aioresponses
from click.testing import CliRunner
from tests import fixtures
from tests import get_jwt
from tests.const import TEST_LOCALE
from tests.const import TEST_LOGIN_TOKEN
from tests.const import TEST_PASSWORD
from tests.const import TEST_PERSON_ID
from tests.const import TEST_USERNAME

from renault_api.cli import __main__
from renault_api.cli.renault_settings import CREDENTIAL_PATH
from renault_api.const import CONF_LOCALE
from renault_api.credential import Credential
from renault_api.credential import JWTCredential
from renault_api.credential_store import FileCredentialStore
from renault_api.gigya import GIGYA_JWT
from renault_api.gigya import GIGYA_LOGIN_TOKEN
from renault_api.gigya import GIGYA_PERSON_ID


def test_login_prompt(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """It exits with a status code of zero."""
    fixtures.inject_gigya_login(mocked_responses)

    result = cli_runner.invoke(
        __main__.main,
        "--debug login",
        input=f"{TEST_USERNAME}\n{TEST_PASSWORD}\n{TEST_LOCALE}\ny",
    )
    assert result.exit_code == 0, result.exception
    expected_output = (
        f"User: {TEST_USERNAME}\n"
        "Password: \n"
        f"Please select a locale [{getdefaultlocale()[0]}]: {TEST_LOCALE}\n"
        "Do you want to save the locale to the credential store? [y/N]: y\n"
        "\n"
    )
    assert expected_output == result.output


def test_login_no_prompt(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """It exits with a status code of zero."""
    fixtures.inject_gigya_login(mocked_responses)

    result = cli_runner.invoke(
        __main__.main,
        f"--debug --locale {TEST_LOCALE} "
        f"login --user {TEST_USERNAME} --password {TEST_PASSWORD}",
    )
    assert result.exit_code == 0, result.exception
    assert "" == result.output


def test_list_accounts_prompt(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    fixtures.inject_gigya_all(mocked_responses)
    fixtures.inject_kamereon_person(mocked_responses)

    result = cli_runner.invoke(
        __main__.main,
        "accounts",
        input=f"{TEST_LOCALE}\nN\n{TEST_USERNAME}\n{TEST_PASSWORD}\n",
    )
    assert result.exit_code == 0, result.exception

    expected_output = (
        f"Please select a locale [{getdefaultlocale()[0]}]: {TEST_LOCALE}\n"
        "Do you want to save the locale to the credential store? [y/N]: N\n"
        "\n"
        f"User: {TEST_USERNAME}\n"
        "Password: \n"
        "\n"
        "Type       ID\n"
        "---------  ------------\n"
        "MYRENAULT  account-id-1\n"
        "SFDC       account-id-2\n"
    )
    assert expected_output == result.output


def test_list_accounts_no_prompt(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    credential_store = FileCredentialStore(os.path.expanduser(CREDENTIAL_PATH))
    credential_store[CONF_LOCALE] = Credential(TEST_LOCALE)
    credential_store[GIGYA_LOGIN_TOKEN] = Credential(TEST_LOGIN_TOKEN)
    credential_store[GIGYA_PERSON_ID] = Credential(TEST_PERSON_ID)
    credential_store[GIGYA_JWT] = JWTCredential(get_jwt())

    fixtures.inject_kamereon_person(mocked_responses)

    result = cli_runner.invoke(__main__.main, "accounts")
    assert result.exit_code == 0, result.exception

    expected_output = (
        "Type       ID\n"
        "---------  ------------\n"
        "MYRENAULT  account-id-1\n"
        "SFDC       account-id-2\n"
    )
    assert expected_output == result.output
