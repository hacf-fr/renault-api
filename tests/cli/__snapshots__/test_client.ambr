# serializer version: 1
# name: test_list_accounts_no_prompt
  '''
  Type       ID
  ---------  ------------
  MYRENAULT  account-id-1
  SFDC       account-id-2
  
  '''
# ---
# name: test_list_accounts_prompt
  '''
  Please select a locale [fr_CA]: fr_FR
  Do you want to save the locale to the credential store? [y/N]: N
  
  User: test@example.com
  Password: 
  
  Type       ID
  ---------  ------------
  MYRENAULT  account-id-1
  SFDC       account-id-2
  
  '''
# ---
# name: test_login_no_prompt
  ''
# ---
# name: test_login_prompt
  '''
  User: test@example.com
  Password: 
  Please select a locale [fr_CA]: fr_FR
  Do you want to save the locale to the credential store? [y/N]: y
  
  
  '''
# ---
